caledonia/src/bounded.rs

```diff
@@ -0,0 +1,451 @@
+    /// Completeness security parameter
+    pub lambda_rel: usize,
+    /// Approximate size of set Sp to lower bound
+    pub n_p: usize,
+    /// Target lower bound
+    pub n_f: usize,
+}
+/// Setup output parameters
+#[derive(Debug, Clone)]
+pub struct Setup {
+    /// Approximate size of set Sp to lower bound
+    pub n_p: usize,
+    /// Proof size (in Sp elements)
+    pub u: usize,
+    /// Proof max counter
+    pub r: usize,
+    /// Proof max 2nd counter
+    pub d: usize,
+    /// Inverse of probability p_q
+    pub q: usize,
+    /// Computation bound
+    pub b: usize,
+}
+impl Setup {
+    /// Setup algorithm taking a Params as input and returning setup parameters (u,d,q)
+    pub fn new(params: &Params) -> Self {
+        // Misc values
+        let e = E;
+        let log_2 = |x: f64| x.log2();
+        let loge = log_2(e);
+        let logloge = log_2(loge);
+        let log3 = log_2(3.0);
+        let log12 = log_2(12.0);
+
+        // Converting params to f64
+        let n_p_f64 = params.n_p as f64;
+        let n_f_f64 = params.n_f as f64;
+        let lognpnf = log_2(n_p_f64 / n_f_f64);
+        let lambda_rel = params.lambda_rel as f64;
+        let lambda_sec = (params.lambda_sec as f64) + log_2(lambda_rel) as f64;
+
+        let l2 = (lambda_rel * lambda_rel).ceil() as usize; // Is it the right lambda?
+        let l3 = l2 * params.lambda_rel;
+
+        // Initialising output params
+        let u;
+        let r;
+        let d;
+        let q;
+        let b;
+
+        // If $n_p$ < λ^3, we define the parameters accoding to Section 3.2.2, Theorem 12 and Corollary 3
+        if n_p_f64 < (l3 as f64) {
+            let u_f64 = (lambda_sec + log_2(lambda_rel) + 5.0 - logloge) / lognpnf;
+            u = u_f64.ceil() as usize;
+            r = params.lambda_rel;
+            b = 1000 * l2; // What about the epsilon
+            let d_f64 = 32.0 * log12 * u_f64;
+            d = d_f64.ceil() as usize;
+            q = ((2.0 * log12) / d_f64).recip().ceil() as usize;
+        } else {
+            // Otherwise, according to Section 3.2, Corrollary 2
+            let u_f64 = (lambda_sec + log_2(lambda_rel + log3) + 1.0 - logloge) / lognpnf;
+            u = u_f64.ceil() as usize;
+            r = 1;
+            b = 1000 * l3;
+            let d_f64 = 16.0 * u_f64 * (lambda_rel + log3) / loge;
+            d = d_f64.ceil() as usize;
+            q = (2.0 * (lambda_rel + log3) / (d_f64 * loge)).recip().ceil() as usize;
+
+            let check = ((d_f64 * d_f64 * loge) / (9.0 * (lambda_rel + log3))).ceil() as usize;
+            assert!(params.n_p >= check);
+        }
+
+        Setup {
+            n_p: params.n_p,
+            u,
+            r,
+            d,
+            q,
+            b,
+        }
+    }
+}
+
+/// Round parameters
+#[derive(Debug, Clone)]
+pub struct Round {
+    /// Proof counter
+    v: usize,
+    /// Proof 2nd counter
+    t: usize,
+    // Round candidate tuple
+    s_list: Vec<[u8; 32]>,
```

Comment:
It would definitely be useful to use types. 

I was thinking of directly using the Hash trait for slist however (https://doc.rust-lang.org/std/hash/trait.Hash.html). 

That way, we wouldn't care about what type of data we are handling as long as it is hashable.

Reply:
Sound definitely like a good idea!

Reply:
Unfortunately this trait only gives the choice of a bad hash so I defined types as you first suggested.